### DISCUSSION TITLE
Generated sha256 key doesnt match github's sha256 key (Webhooks).

### DISCUSSION THREAD

**Discussion by flojo007** [community/community#28401] [Apps, API and Webhooks]
Hello,

When I generate my sha256 key, I always get a different one than the one github's adds to the headers with the webhooks.

My code to generate it:
`app.post('/github-push', async function (req, res) {
console.log("sha256=" + crypto.createHmac('sha256', process.env.GITHUB_PUSH_KEY + req.read).update("json").digest("hex"))
}`

Do you have any ideas?

Thanks, flojo

**Comment by ghost** [comment]
Bumping this instead of making a new discussion. 

I am also having trouble validating hashes properly. I am working with python and using the flask framework for my server. 
```python
@app.route("/github_webhook", methods=["POST"])
def github_webhook():
    if "X-Hub-Signature-256" in request.headers:
        compute_hash = hmac.new(github_secret, request.data, digestmod="sha256").hexdigest()
        if hmac.compare_digest(request.headers["X-Hub-Signature-256"], f"sha256={compute_hash}"):
            return "webhook hash validated!", 202
    return "webhook hash does not match!", 401
```

This code always results in a failure to match the hashes from actual webhooks sent by github. Testing locally and using postman to send the POST requests, i managed to get this to work IF i sent the request using the hash and json copy and pasted from the real github request except put all the json on a single line and removed all the white space between arrays in the json body (so basically minifying the json payload)

But the github webhooks recent deliveries tab shows the payload as pretty printed. Is this actually what the payload looks like when github creates the hash? there definitely seems to be a difference with how github is sending the payload, and how flask creates the `request.data` object. What is the correct way to match the hashes for the payloads?

**Comment by Morriz** [comment]
bumpppp

**Comment by codebeast0420** [comment]
When calculating the SHA256 HMAC (Hash-Based Message Authentication Code) for GitHub webhook payloads, it's important to ensure that you're using the correct key and data to produce the hash. Your current code snippet has a few issues that might be causing the discrepancy between your calculated HMAC and the one provided by GitHub.
